### ISSUE TITLE
Implement SA1139: Use literal suffix notation instead of casting

### ISSUE THREAD

**Issue by berkeleybross** [DotNetAnalyzers/StyleCopAnalyzers#1562]
New rule proposal: Use literal suffix notation instead of casting

_Category_: Readability
_Rule_: When a numeric literal is immediately cast to another numeric type, it should be replaced with a literal suffix.
_Rationale_: Improves readability of code, and (presumably) run time casts.

Bad:

``` CSharp
long foo = (long)50;
```

Good:

``` CSharp
long foo = 50L;
```

See https://msdn.microsoft.com/en-us/library/aa691085(v=vs.71).aspx and https://msdn.microsoft.com/en-us/library/aa664674(v=vs.71).aspx for the specification for integer/real literals.

Things that might need considering:

The literal suffix can be upper and lower case. I would prefer uppercase, (and if you read the integer literal specification above, it recommends this).

This should work on nullable types as well:

``` CSharp
long? foo = (long?)50;
```

Should be fixed to 

``` CSharp
long? foo = 50L;
```

**Comment by vweijsters** [comment]
:+1:

**Comment by nbarbettini** [comment]
:+1:

**Comment by GregReddick** [comment]
:+1: See #1563, though, on the capitalization of the literal specification.

**Comment by otac0n** [comment]
:+1:

**Comment by sharwell** [comment]
:memo: Make sure to not report a warning on the following code:

```
var foo = (long?)50;
```

**Comment by tmaczynski** [comment]
I've decided to make some contributions to DotNetAnalyzers. I've done a draft of analyzer and tests and I'll implement code-fix later.

I've noticed that in some cases (e.g. when a field `long x = (long)1` is declared) this analysis is overlapping with `IDE0004 C# Cast is redundant`. Codefix for IDE0004 deletes cast and relies on an explicit casting. 

I think that SA1139 should be reported in such case and another codefix should be provided (which creates a literal of a desired type) but I wanted to ask about your thoughts before implementing it.

I'll do a pull request when I'm finished with this analyzer, my [first implementation](https://github.com/tmaczynski/StyleCopAnalyzers/compare/8483314b3282b831ae1feba3f4070b0d0e853de1...tmaczynski:f08086ba8ebf699ec86a847e00f800ce40cb075e) is in my forked repository.

**Comment by tmaczynski** [comment]
I was extremely busy recently with my full time job, but I’ve found some time to implement first draft that is worth reviewing and (hopefully) checking in into main repo.

I implemented analysis that takes into account all integer and real literals (`ints`, `uints`, `longs`, `ulongs`, `floats`, `doubles` and `decimals`). My implementation has appropriate unit tests.

In my code I took into account:

- Differences in values between checked and unchecked code
- Original format of a number in majority of cases (e.g. `(float)2.5` becomes `2.5f`, `(float)2e5` becomes `2e5f` and `(long)0x5AA` becomes `0x5AAL`)
- Correct rounding when a real number is casted on integer type

My analysis is reported only for type aliases (e.g. `int`, `float`), not for full type names (e.g. `Int32`, `Single`). If full type name is used SA1121 is reported. Also, this analysis isn’t reported when a cast is invalid (e.g. `(uint)-1` in checked environment) – in that case a compiler error is produced. 

IMHO, nullable types should be covered by another analyzer. Nullable types have no literals – programmer must use casts, rely on explicit conversions or use `new Nullable<>()` constructor.  I don't think that relying on explicit conversions automatically makes code clearer (especially if literals are arguments of a method). 

I don’t know whether contributors are listed somewhere. I added comments listing me as a contributor in my pull request, but feel free to remove them if you want.
